### ISSUE TITLE
ssh keys at salt:// urls are unusable with --local due to open permissions in cache

### ISSUE THREAD

**Issue by guest20** [saltstack/salt#44325]
### Description of Issue/Question

This looks a lot like #38144, except that none of the sneaky "hide the key on the master" tricks mentioned in the tread help because I'm running [masterless](https://docs.saltstack.com/en/latest/topics/tutorials/quickstart.html).

When referencing a private key served from `gitfs`, in a `--local` run, the permissions on the cache are too open, and so git/ssh refuse to use it with:
```
WARNING: UNPROTECTED PRIVATE KEY FILE! Permissions 0644 for '/var/cache/salt/minion/files/base/webserver/host/default-site/deploy-key' are too open.
```

### Setup

in some `.sls` that applies to this computer:
``` yaml
/websites/default/default-site:
  git.latest:
    - name: ssh://git@gitlab.com/guest20/default-site.git
    - target: /websites/default/default-site
    - identity: salt://webserver/host/default-site/deploy-key
```

my minion config:
```
        #masterless-minion (from cloud-init)
        master: []
        grains: { local_etc: /etc/, etc: /etc/ }
        state_top_saltenv: base
        fileserver_backend: [git]
        gitfs_remotes:
          - git@gitlab.com:guest20/salt.git:
             - root: states
```

### Steps to Reproduce Issue

1. one git repo with salt states is `salt.git`. The key for this repo is in `/home/.ssh/id_rsa` (dropped by cloud-init)
1. a second git repo with "welcome to your webserver" is `default-site.git`. The key for this repo is in `salt.git`
1. reference the key in `salt.git` from the `identity` key of a `git.latest` via a `salt://` url
1. apply the states with `salt-call --local state.apply`

```
WARNING: UNPROTECTED PRIVATE KEY FILE! Permissions 0644 for '/var/cache/salt/minion/files/base/webserver/host/default-site/deploy-key' are too open.
```
### disclaimers

1. yes, I have my websites in `/websites/`, it's odd but less of a pain when moving to/from bsd/linux.
1. yes, I should use a better way to store my deploy keys, but they only have read access to the repos.
1. no, I'm not really sure how I could keep anything secret on a masterless setup, I suspect it might involve [External Pillars](https://docs.saltstack.com/en/latest/topics/development/external_pillars.html#external-pillars) and GPG?.

At this point anything I drop on the fs via `cloud-init` is completely open to any user on the machine, and so are my secrets in `salt.git`, so something that addresses both would be ideal.

### Versions Report

```
# salt-call --versions-report
Salt Version:
           Salt: 2016.11.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.13 (default, Jan 19 2017, 14:48:08)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.2
        machine: x86_64
        release: 4.9.0-4-amd64
         system: Linux
        version: debian 9.2
```

### Work 'round

I ended up replacing the `salt://` url with a `file.managed` with narrower permissions then referring to that by its local path, but I can't shake the feeling that key is still visible in `/var/cache` .

Would it be better if the permissions on the cache were narrower, over all?

**Comment by garethgreenaway** [comment]
@guest20 Thanks for the report.  Definitely not the ideal case, but I don't believe we've seen reports of it happening without `--local` so need to dig in a bit.

**Comment by gtmanfred** [comment]
I believe that this is fixed in a later version of 2016.11,

**Comment by gtmanfred** [comment]
This should be fixed by https://github.com/saltstack/salt/pull/40609 Which was included in 2016.11.4

**Comment by guest20** [comment]
I had a look at #40609 and it looks like it makes mode changes in the git-managed cache dir.

I'm not sure how the git in salt works, but I have a feeling that each git operation will "fix" those permissions back to the 777 / 666 that git prefers (if it doesn't bail out because it looks like a local change).

I'll give it a go now and see how much magic is in the new version.
